sdk/core/azure-core/src/http/curl/curl_connection_pool_private.hpp

```diff
@@ -90,51 +87,47 @@ namespace Azure { namespace Core { namespace Http { namespace _detail {
+    void MoveConnectionBackToPool(
         std::unique_ptr<CurlNetworkConnection> connection,
         HttpStatusCode lastStatusCode);
 
-    // Class can't have instances.
-    CurlConnectionPool() = delete;
-
-    static void StopCleaner() { g_isCleanConnectionsRunning = false; }
+    /**
+     * @brief Remove all connections from the pool.
+     *
+     */
+    void ResetPool() { ConnectionPoolIndex.clear(); }
 
-  private:
     /**
-     * Review all connections in the pool and removes old connections that might be already
-     * expired and closed its connection on server side.
+     * @brief Keeps a unique key for each host and creates a connection pool for each key.
+     *
+     * @details This way getting a connection for a specific host can be done in O(1) instead of
+     * looping a single connection list to find the first connection for the required host.
+     *
+     * @remark There might be multiple connections for each host.
      */
-    static void CleanUp();
+    std::map<std::string, std::list<std::unique_ptr<CurlNetworkConnection>>> ConnectionPoolIndex;
 
-    AZ_CORE_DLLEXPORT static uint64_t g_connectionCounter;
-    AZ_CORE_DLLEXPORT static std::atomic<bool> g_isCleanConnectionsRunning;
-    // Removes all connections and indexes
-    static void ClearIndex() { CurlConnectionPool::ConnectionPoolIndex.clear(); }
+    std::mutex ConnectionPoolMutex;
 
-    // Makes possible to know the number of current connections in the connection pool for an
-    // index
-    static int64_t ConnectionsOnPool(std::string const& host)
-    {
-      auto& pool = CurlConnectionPool::ConnectionPoolIndex[host];
-      return pool.size();
-    };
+    // This is used to put the cleaning pool thread to sleep and yet to be able to wake it if the
+    // application finishes.
+    std::condition_variable ConditionalVariableForCleanThread;
 
-    // Makes possible to know the number indexes in the pool
-    static int64_t ConnectionsIndexOnPool()
-    {
-      return CurlConnectionPool::ConnectionPoolIndex.size();
-    };
-  };
+    AZ_CORE_DLLEXPORT static Azure::Core::Http::_detail::CurlConnectionPool g_curlConnectionPool;
 
-  struct CurlGlobalStateForAzureSdk
-  {
-    CurlGlobalStateForAzureSdk() { curl_global_init(CURL_GLOBAL_ALL); }
+  private:
+    // private constructor to keep this as singleton.
+    CurlConnectionPool() { curl_global_init(CURL_GLOBAL_ALL); }
 
-    ~CurlGlobalStateForAzureSdk()
-    {
-      CurlConnectionPool::StopCleaner();
-      curl_global_cleanup();
-    }
+    std::thread m_cleanThread;
+
+    // Makes possible to know the number of current connections in the connection pool for an
+    // index
+    int64_t ConnectionsOnPool(std::string const& host) { return ConnectionPoolIndex[host].size(); };
+
+    // Define the maximun allowed connections per host-index in the pool. If this number is reached
+    // for the host-index, next connections trying to be added to the pool will be ignored.
```

Comment:
```suggestion
    // for the host-index, new connections attempted to be added to the pool will be ignored.
```

Reply:
Instead of them being ignored, should the oldest connection in the pool get replaced by it?

Reply:
> Instead of them being ignored, should the oldest connection in the pool get replaced by it?

That's a good idea, yes

Reply:
To be clear, to maintain the LIFO/stack behavior, basically you erase the bottom item, and push this newest item at the top (and not at the bottom).

It's essentially a bounded stack, but we implement it as a list under the covers because we want to be able to remove from either end in O(1).

Otherwise, it's not purely LIFO, which is OK too if you want to go down that route for some reason.